### ISSUE TITLE
Sandcastle build seems to keep filehandle open, preventing successive builds (VS restart required)

### ISSUE THREAD

**Issue by RalfKoban** [EWSoftware/SHFB#17]
Hello,

I'm using the Sandcastle help file builder to build my API documentation file each time I trigger a release build.

It seems that at least in the latest version (2015.01.12) the filehandle(s) to my DLL(s) are not closed anymore after build. That prevents me from triggering new builds completely; I need to restart VS (2013) instead each time.
If I don't build the API documentation, I'm able to trigger release build by release build without any problem.

Best regards,
Ralf Koban

**Comment by EWSoftware** [comment]
VS 2013 supports parallel builds.  If you haven't set the project dependencies, it may be trying to build the projects out of order thus you're seeing locked assemblies as it's trying to build the SHFB project at the same time as the assembly project.  Set the project dependencies such that the SHFB project is dependent on the assembly projects.  When done that way, the SHFB project won't build in parallel with the assembly projects and it should resolve the issue.

**Comment by RalfKoban** [comment]
Hello,
The project dependencies are set the way that my SHFB project depends on all the other projects. So it's built at the very end of the build chain.
If I take a look into the file handles that are consumed by devenv via the SysInternals ProcessExplorer, then I see that the file handle(s) to my DLL(s) are created when the SHFB project gets built but they are not released anymore. If I don't build the SHFB project, then I don't see those file handle(s).
So I suppose that they are created either by SHFB directly or by one of the chosen SHFB build components (I'm using API Token Resolution, Code Block Component and Intellisense Component).

**Comment by EWSoftware** [comment]
MRefBuilder is the only tool that would be accessing the assemblies.  It runs as a separate task and, once it's done, it shuts down and I can't see how it would be hanging on to the assemblies.  I'm unable to duplicate the problem with any of my own projects or the test case in SHFB.  Are you able to put together an example that demonstrates the problem?

**Comment by EWSoftware** [comment]
I ran into this issue in one of my projects but it didn't contain a SHFB project so I think it may not be related to SHFB at all.  Searching Google turns up quite a few cases where VS 2013 is locking the assemblies.  In my case, I ran a Clean to remove the output which it did successfully and then rebuilt the project and it worked as usual.  You might give that a try.

**Comment by EWSoftware** [comment]
I've figured out what's going on.  Somebody else reported this issue via e-mail with a reference to a project that was showing the problem.  If your project is like that one, the SHFB project is in the same folder as the solution or the projects being documented.

SHFB always searches the current project folder and all subfolders for components it can use for presentation styles, build components, plug-ins etc.  Unfortunately, MEF does not release assemblies even if they don't contain anything of use and thus they remain loaded in the app domain until the host exits.

The solution is simple.  Move your SHFB project to a subfolder such as .\Doc or .\Documentation so that when it goes looking for components it won't find the assemblies being documented as they are outside the SHFB project's folder.  You can run into a similar problem if the project's Component Path project property is set to a folder where it can find the documented assemblies too.  Once the project is moved, you may need to adjust the paths to the documentation sources so that it can find them.

Since I always put my SHFB projects in a separate folder, I wasn't running into the issue.

**Comment by RalfKoban** [comment]
Thanks for the update on this, copying the SHFB project into a separate folder did the trick. :)
